src/stream/src/executor/aggregation/approx_count_distinct.rs

```diff
@@ -27,94 +27,155 @@ use risingwave_common::types::{Datum, DatumRef, Scalar, ScalarImpl};
+    }
+
+    fn last_key(&self) -> u8 {
+        assert!(!self.is_empty());
+        self.inner.last().unwrap().0
+    }
+}
 
-        if index >= 33 {
-            return Ok((self.count_33_to_64 >> (index - 33)) & 1);
+#[derive(Clone, Debug)]
+struct RegisterBucket<const DENSE_BITS: usize> {
+    dense_counts: [u64; DENSE_BITS],
+    sparse_counts: Option<SparseCount>,
+}
+
+impl<const DENSE_BITS: usize> RegisterBucket<DENSE_BITS> {
+    pub fn new() -> Self {
+        Self {
+            dense_counts: [0u64; DENSE_BITS],
+            sparse_counts: None,
         }
+    }
 
-        if index >= 25 {
-            return Ok(self.count_25_to_32[index - 25] as u32);
+    fn get_bucket(&self, index: usize) -> StreamExecutorResult<u64> {
+        if index > 64 || index == 0 {
+            bail!("HyperLogLog: Invalid bucket index");
         }
 
-        if index >= 17 {
-            return Ok(self.count_17_to_24[index - 17] as u32);
+        if index > DENSE_BITS {
+            if let Some(counts) = &self.sparse_counts {
+                return Ok(counts.get(index as u8));
+            } else {
+                return Ok(0);
+            }
         }
 
-        Ok(self.count_1_to_16[index - 1])
+        Ok(self.dense_counts[index - 1])
     }
 
     /// Increments or decrements the bucket at `index` depending on the state of `is_insert`.
     /// Returns an Error if `index` is invalid or if inserting will cause an overflow in the bucket.
     fn update_bucket(&mut self, index: usize, is_insert: bool) -> StreamExecutorResult<()> {
-        if !(1..=64).contains(&index) {
+        if index > 64 || index == 0 {
             bail!("HyperLogLog: Invalid bucket index");
         }
 
         let count = self.get_bucket(index)?;
 
         if is_insert {
-            if index >= 33 {
-                if count == 1 {
-                    bail!("HyperLogLog: Count exceeds maximum bucket value. Your data stream may have too many repeated values or too large a cardinality for approx_count_distinct to handle.");
-                }
-                self.count_33_to_64 |= 1 << (index - 33);
-            } else if index >= 25 {
-                if count as u8 == u8::MAX {
-                    bail!("HyperLogLog: Count exceeds maximum bucket value. Your data stream may have too many repeated values or too large a cardinality for approx_count_distinct to handle.");
-                }
-                self.count_25_to_32[index - 25] = count as u8 + 1;
-            } else if index >= 17 {
-                if count as u16 == u16::MAX {
-                    bail!("HyperLogLog: Count exceeds maximum bucket value. Your data stream may have too many repeated values or too large a cardinality for approx_count_distinct to handle.");
+            if index > DENSE_BITS {
+                if let Some(counts) = &mut self.sparse_counts {
+                    counts.add(index as u8);
+                } else {
+                    let mut counts = SparseCount::new();
```

Comment:
Correct me if I am wrong:
if `index` > `DENSE_BITS`, i.e. the number of trailing zeros is very large, it implies we are entering a `high cardinality` scenario, then we switch from `sparse` to `dense` instead of the other way around?

As shown in https://github.com/redis/redis/blob/unstable/src/hyperloglog.c#L663, `if (count > HLL_SPARSE_VAL_MAX_VALUE)` happens, then Redis promote it from sparse representation to dense representation.

Reply:
So we are not implementing a sparse-dense transition. 

We are implementing a hybrid representation where we don't pay the memory cost for the sparsely-represented higher bits up front, at the cost of a bit more compute for the updates.

Reply:
I see👌